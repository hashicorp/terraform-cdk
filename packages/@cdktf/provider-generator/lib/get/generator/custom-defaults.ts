// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
export const CUSTOM_DEFAULTS: { [fullName: string]: any } = {
  "aws.security_group.ingress.cidr_blocks": null,
  "aws.security_group.ingress.description": null,
  "aws.security_group.ingress.from_port": null,
  "aws.security_group.ingress.ipv6_cidr_blocks": null,
  "aws.security_group.ingress.prefix_list_ids": null,
  "aws.security_group.ingress.protocol": null,
  "aws.security_group.ingress.security_groups": null,
  "aws.security_group.ingress.self": null,
  "aws.security_group.ingress.to_port": null,
  "aws.security_group.egress.cidr_blocks": null,
  "aws.security_group.egress.description": null,
  "aws.security_group.egress.from_port": null,
  "aws.security_group.egress.ipv6_cidr_blocks": null,
  "aws.security_group.egress.prefix_list_ids": null,
  "aws.security_group.egress.protocol": null,
  "aws.security_group.egress.security_groups": null,
  "aws.security_group.egress.self": null,
  "aws.security_group.egress.to_port": null,

  "aws.default_route_table.route.cidr_block": null,
  "aws.default_route_table.route.ipv6_cidr_block": null,
  "aws.default_route_table.route.egress_only_gateway_id": null,
  "aws.default_route_table.route.gateway_id": null,
  "aws.default_route_table.route.instance_id": null,
  "aws.default_route_table.route.nat_gateway_id": null,
  "aws.default_route_table.route.network_interface_id": null,
  "aws.default_route_table.route.transit_gateway_id": null,
  "aws.default_route_table.route.vpc_endpoint_id": null,
  "aws.default_route_table.route.vpc_peering_connection_id": null,
  "aws.default_route_table.route.destination_prefix_list_id": null,
  "aws.default_route_table.route.core_network_arn": null,

  "aws.emr_cluster.step.action_on_failure": null,
  "aws.emr_cluster.step.hadoop_jar_step": null,
  "aws.emr_cluster.step.name": null,
  "aws.emr_cluster.step.hadoop_jar_step.args": null,
  "aws.emr_cluster.step.hadoop_jar_step.jar": null,
  "aws.emr_cluster.step.hadoop_jar_step.main_class": null,
  "aws.emr_cluster.step.hadoop_jar_step.properties": null,

  "aws.network_acl.ingress.from_port": null,
  "aws.network_acl.ingress.to_port": null,
  "aws.network_acl.ingress.rule_no": null,
  "aws.network_acl.ingress.action": null,
  "aws.network_acl.ingress.protocol": null,
  "aws.network_acl.ingress.cidr_block": null,
  "aws.network_acl.ingress.ipv6_cidr_block": null,
  "aws.network_acl.ingress.icmp_type": null,
  "aws.network_acl.ingress.icmp_code": null,
  "aws.network_acl.egress.from_port": null,
  "aws.network_acl.egress.to_port": null,
  "aws.network_acl.egress.rule_no": null,
  "aws.network_acl.egress.action": null,
  "aws.network_acl.egress.protocol": null,
  "aws.network_acl.egress.cidr_block": null,
  "aws.network_acl.egress.ipv6_cidr_block": null,
  "aws.network_acl.egress.icmp_type": null,
  "aws.network_acl.egress.icmp_code": null,

  "aws.route_table.route.cidr_block": null,
  "aws.route_table.route.ipv6_cidr_block": null,
  "aws.route_table.route.egress_only_gateway_id": null,
  "aws.route_table.route.gateway_id": null,
  "aws.route_table.route.instance_id": null,
  "aws.route_table.route.nat_gateway_id": null,
  "aws.route_table.route.local_gateway_id": null,
  "aws.route_table.route.network_interface_id": null,
  "aws.route_table.route.transit_gateway_id": null,
  "aws.route_table.route.vpc_endpoint_id": null,
  "aws.route_table.route.vpc_peering_connection_id": null,
  "aws.route_table.route.destination_prefix_list_id": null,
  "aws.route_table.route.carrier_gateway_id": null,
  "aws.route_table.route.core_network_arn": null,

  "aws.default_security_group.ingress.cidr_blocks": null,
  "aws.default_security_group.ingress.description": null,
  "aws.default_security_group.ingress.from_port": null,
  "aws.default_security_group.ingress.ipv6_cidr_blocks": null,
  "aws.default_security_group.ingress.prefix_list_ids": null,
  "aws.default_security_group.ingress.protocol": null,
  "aws.default_security_group.ingress.security_groups": null,
  "aws.default_security_group.ingress.self": null,
  "aws.default_security_group.ingress.to_port": null,
  "aws.default_security_group.egress.cidr_blocks": null,
  "aws.default_security_group.egress.description": null,
  "aws.default_security_group.egress.from_port": null,
  "aws.default_security_group.egress.ipv6_cidr_blocks": null,
  "aws.default_security_group.egress.prefix_list_ids": null,
  "aws.default_security_group.egress.protocol": null,
  "aws.default_security_group.egress.security_groups": null,
  "aws.default_security_group.egress.self": null,
  "aws.default_security_group.egress.to_port": null,

  "azuread.application.app_role.id": null,
  "azuread.application.app_role.allowed_member_types": null,
  "azuread.application.app_role.description": null,
  "azuread.application.app_role.display_name": null,
  "azuread.application.app_role.is_enabled": null,
  "azuread.application.app_role.value": null,
  "azuread.application.oauth2_permissions.id": null,
  "azuread.application.oauth2_permissions.admin_consent_description": null,
  "azuread.application.oauth2_permissions.admin_consent_display_name": null,
  "azuread.application.oauth2_permissions.value": null,
  "azuread.application.oauth2_permissions.type": null,
  "azuread.application.oauth2_permissions.is_enabled": null,
  "azuread.application.oauth2_permissions.user_consent_description": null,
  "azuread.application.oauth2_permissions.user_consent_display_name": null,

  "azurestack.route_table.route.name": null,
  "azurestack.route_table.route.address_prefix": null,
  "azurestack.route_table.route.next_hop_type": null,
  "azurestack.route_table.route.next_hop_in_ip_address": null,

  "vault.auth_backend.tune.default_lease_ttl": null,
  "vault.auth_backend.tune.max_lease_ttl": null,
  "vault.auth_backend.tune.audit_non_hmac_request_keys": null,
  "vault.auth_backend.tune.audit_non_hmac_response_keys": null,
  "vault.auth_backend.tune.listing_visibility": null,
  "vault.auth_backend.tune.passthrough_request_headers": null,
  "vault.auth_backend.tune.allowed_response_headers": null,
  "vault.auth_backend.tune.token_type": null,

  "vault.github_auth_backend.tune.default_lease_ttl": null,
  "vault.github_auth_backend.tune.max_lease_ttl": null,
  "vault.github_auth_backend.tune.audit_non_hmac_request_keys": null,
  "vault.github_auth_backend.tune.audit_non_hmac_response_keys": null,
  "vault.github_auth_backend.tune.listing_visibility": null,
  "vault.github_auth_backend.tune.passthrough_request_headers": null,
  "vault.github_auth_backend.tune.allowed_response_headers": null,
  "vault.github_auth_backend.tune.token_type": null,

  "vault.jwt_auth_backend.tune.default_lease_ttl": null,
  "vault.jwt_auth_backend.tune.max_lease_ttl": null,
  "vault.jwt_auth_backend.tune.audit_non_hmac_request_keys": null,
  "vault.jwt_auth_backend.tune.audit_non_hmac_response_keys": null,
  "vault.jwt_auth_backend.tune.listing_visibility": null,
  "vault.jwt_auth_backend.tune.passthrough_request_headers": null,
  "vault.jwt_auth_backend.tune.allowed_response_headers": null,
  "vault.jwt_auth_backend.tune.token_type": null,

  "vault.okta_auth_backend.group.group_name": null,
  "vault.okta_auth_backend.group.policies": null,
  "vault.okta_auth_backend.user.username": null,
  "vault.okta_auth_backend.user.groups": null,
  "vault.okta_auth_backend.user.policies": null,

  "google.compute_instance.guest_accelerator.count": null,
  "google.compute_instance.guest_accelerator.type": null,

  "google.compute_instance_from_template.attached_disk.source": null,
  "google.compute_instance_from_template.attached_disk.device_name": null,
  "google.compute_instance_from_template.attached_disk.mode": null,
  "google.compute_instance_from_template.attached_disk.disk_encryption_key_raw":
    null,
  "google.compute_instance_from_template.attached_disk.kms_key_self_link": null,
  "google.compute_instance_from_template.guest_accelerator.count": null,
  "google.compute_instance_from_template.guest_accelerator.type": null,
  "google.compute_instance_from_template.service_account.email": null,
  "google.compute_instance_from_template.service_account.scopes": null,
  "google.compute_instance_from_template.scratch_disk.interface": null,
  "google.compute_instance_from_template.network_interface.alias_ip_range.ip_cidr_range":
    null,
  "google.compute_instance_from_template.network_interface.alias_ip_range.subnetwork_range_name":
    null,
  "google.compute_instance_from_template.network_interface.access_config.nat_ip":
    null,
  "google.compute_instance_from_template.network_interface.access_config.public_ptr_domain_name":
    null,
  "google.compute_instance_from_template.network_interface.access_config.network_tier":
    null,

  "google.compute_subnetwork.secondary_ip_range.range_name": null,
  "google.compute_subnetwork.secondary_ip_range.ip_cidr_range": null,

  "google.container_cluster.node_config.guest_accelerator.count": null,
  "google.container_cluster.node_config.guest_accelerator.type": null,
  "google.container_cluster.node_config.taint.key": null,
  "google.container_cluster.node_config.taint.value": null,
  "google.container_cluster.node_config.taint.effect": null,

  "google.container_node_pool.node_config.guest_accelerator.count": null,
  "google.container_node_pool.node_config.guest_accelerator.type": null,
  "google.container_node_pool.node_config.taint.key": null,
  "google.container_node_pool.node_config.taint.value": null,
  "google.container_node_pool.node_config.taint.effect": null,

  "google.composer_environment.ip_allocation_policy.use_ip_aliases": null,
  "google.composer_environment.ip_allocation_policy.cluster_secondary_range_name":
    null,
  "google.composer_environment.ip_allocation_policy.services_secondary_range_name":
    null,
  "google.composer_environment.ip_allocation_policy.cluster_ipv4_cidr_block":
    null,
  "google.composer_environment.ip_allocation_policy.services_ipv4_cidr_block":
    null,

  "azurerm.eventgrid_topic.inbound_ip_rule.ip_mask": null,
  "azurerm.eventgrid_topic.inbound_ip_rule.action": null,

  "azurerm.eventgrid_domain.inbound_ip_rule.ip_mask": null,
  "azurerm.eventgrid_domain.inbound_ip_rule.action": null,

  "azurerm.container_registry.network_rule_set.default_action": null,
  "azurerm.container_registry.network_rule_set.ip_rule": null,
  "azurerm.container_registry.network_rule_set.virtual_network": null,
  "azurerm.container_registry.network_rule_set.ip_rule.action": null,
  "azurerm.container_registry.network_rule_set.ip_rule.ip_range": null,
  "azurerm.container_registry.network_rule_set.virtual_network.action": null,
  "azurerm.container_registry.network_rule_set.virtual_network.subnet_id": null,
  "azurerm.container_registry.retention_policy.days": null,
  "azurerm.container_registry.retention_policy.enabled": null,
  "azurerm.container_registry.trust_policy.enabled": null,

  "azurerm.eventhub_namespace.network_rulesets.default_action": null,
  "azurerm.eventhub_namespace.network_rulesets.trusted_service_access_enabled":
    null,
  "azurerm.eventhub_namespace.network_rulesets.virtual_network_rule": null,
  "azurerm.eventhub_namespace.network_rulesets.virtual_network_rule.subnet_id":
    null,
  "azurerm.eventhub_namespace.network_rulesets.virtual_network_rule.ignore_missing_virtual_network_service_endpoint":
    null,
  "azurerm.eventhub_namespace.network_rulesets.ip_rule": null,
  "azurerm.eventhub_namespace.network_rulesets.ip_rule.ip_mask": null,
  "azurerm.eventhub_namespace.network_rulesets.ip_rule.action": null,

  "azurerm.automation_runbook.job_schedule.schedule_name": null,
  "azurerm.automation_runbook.job_schedule.parameters": null,
  "azurerm.automation_runbook.job_schedule.run_on": null,
  "azurerm.automation_runbook.job_schedule.job_schedule_id": null,

  "azurerm.mssql_database.extended_auditing_policy.storage_account_access_key":
    null,
  "azurerm.mssql_database.extended_auditing_policy.storage_endpoint": null,
  "azurerm.mssql_database.extended_auditing_policy.storage_account_access_key_is_secondary":
    null,
  "azurerm.mssql_database.extended_auditing_policy.retention_in_days": null,

  "azurerm.mssql_server.extended_auditing_policy.storage_account_access_key":
    null,
  "azurerm.mssql_server.extended_auditing_policy.storage_endpoint": null,
  "azurerm.mssql_server.extended_auditing_policy.storage_account_access_key_is_secondary":
    null,
  "azurerm.mssql_server.extended_auditing_policy.retention_in_days": null,

  "azurerm.route_filter.rule.access": null,
  "azurerm.route_filter.rule.communities": null,
  "azurerm.route_filter.rule.name": null,
  "azurerm.route_filter.rule.rule_type": null,

  "azurerm.route_table.route.name": null,
  "azurerm.route_table.route.address_prefix": null,
  "azurerm.route_table.route.next_hop_type": null,
  "azurerm.route_table.route.next_hop_in_ip_address": null,

  "azurerm.site_recovery_replicated_vm.managed_disk.disk_id": null,
  "azurerm.site_recovery_replicated_vm.managed_disk.staging_storage_account_id":
    null,
  "azurerm.site_recovery_replicated_vm.managed_disk.target_resource_group_id":
    null,
  "azurerm.site_recovery_replicated_vm.managed_disk.target_disk_type": null,
  "azurerm.site_recovery_replicated_vm.managed_disk.target_replica_disk_type":
    null,
  "azurerm.site_recovery_replicated_vm.network_interface.source_network_interface_id":
    null,
  "azurerm.site_recovery_replicated_vm.network_interface.target_static_ip":
    null,
  "azurerm.site_recovery_replicated_vm.network_interface.target_subnet_name":
    null,

  "azurerm.iothub.endpoint.type": null,
  "azurerm.iothub.endpoint.connection_string": null,
  "azurerm.iothub.endpoint.name": null,
  "azurerm.iothub.endpoint.batch_frequency_in_seconds": null,
  "azurerm.iothub.endpoint.max_chunk_size_in_bytes": null,
  "azurerm.iothub.endpoint.container_name": null,
  "azurerm.iothub.endpoint.encoding": null,
  "azurerm.iothub.endpoint.file_name_format": null,
  "azurerm.iothub.endpoint.resource_group_name": null,
  "azurerm.iothub.route.name": null,
  "azurerm.iothub.route.source": null,
  "azurerm.iothub.route.condition": null,
  "azurerm.iothub.route.endpoint_names": null,
  "azurerm.iothub.route.enabled": null,
  "azurerm.iothub.enrichment.key": null,
  "azurerm.iothub.enrichment.value": null,
  "azurerm.iothub.enrichment.endpoint_names": null,

  "azurerm.virtual_network.subnet.name": null,
  "azurerm.virtual_network.subnet.address_prefix": null,
  "azurerm.virtual_network.subnet.security_group": null,
  "azurerm.virtual_network.subnet.id": null,

  "azurerm.network_security_group.security_rule.name": null,
  "azurerm.network_security_group.security_rule.description": null,
  "azurerm.network_security_group.security_rule.protocol": null,
  "azurerm.network_security_group.security_rule.source_port_range": null,
  "azurerm.network_security_group.security_rule.source_port_ranges": null,
  "azurerm.network_security_group.security_rule.destination_port_range": null,
  "azurerm.network_security_group.security_rule.destination_port_ranges": null,
  "azurerm.network_security_group.security_rule.source_address_prefix": null,
  "azurerm.network_security_group.security_rule.source_address_prefixes": null,
  "azurerm.network_security_group.security_rule.destination_address_prefix":
    null,
  "azurerm.network_security_group.security_rule.destination_address_prefixes":
    null,
  "azurerm.network_security_group.security_rule.destination_application_security_group_ids":
    null,
  "azurerm.network_security_group.security_rule.source_application_security_group_ids":
    null,
  "azurerm.network_security_group.security_rule.access": null,
  "azurerm.network_security_group.security_rule.priority": null,
  "azurerm.network_security_group.security_rule.direction": null,

  "azurerm.synapse_workspace.aad_admin.login": null,
  "azurerm.synapse_workspace.aad_admin.object_id": null,
  "azurerm.synapse_workspace.aad_admin.tenant_id": null,

  "azurerm.batch_pool.container_configuration.container_registries.registry_server":
    null,
  "azurerm.batch_pool.container_configuration.container_registries.user_name":
    null,
  "azurerm.batch_pool.container_configuration.container_registries.password":
    null,

  "azurerm.key_vault.access_policy.tenant_id": null,
  "azurerm.key_vault.access_policy.object_id": null,
  "azurerm.key_vault.access_policy.application_id": null,
  "azurerm.key_vault.access_policy.certificate_permissions": null,
  "azurerm.key_vault.access_policy.key_permissions": null,
  "azurerm.key_vault.access_policy.secret_permissions": null,
  "azurerm.key_vault.access_policy.storage_permissions": null,

  "azurerm.app_service.ip_restriction.ip_address": null,
  "azurerm.app_service.ip_restriction.service_tag": null,
  "azurerm.app_service.ip_restriction.subnet_id": null,
  "azurerm.app_service.ip_restriction.virtual_network_subnet_id": null,
  "azurerm.app_service.ip_restriction.name": null,
  "azurerm.app_service.ip_restriction.priority": null,
  "azurerm.app_service.ip_restriction.action": null,
  "azurerm.app_service.scm_ip_restriction.ip_address": null,
  "azurerm.app_service.scm_ip_restriction.service_tag": null,
  "azurerm.app_service.scm_ip_restriction.subnet_id": null,
  "azurerm.app_service.scm_ip_restriction.virtual_network_subnet_id": null,
  "azurerm.app_service.scm_ip_restriction.name": null,
  "azurerm.app_service.scm_ip_restriction.priority": null,
  "azurerm.app_service.scm_ip_restriction.action": null,

  "azurerm.function_app.ip_restriction.ip_address": null,
  "azurerm.function_app.ip_restriction.service_tag": null,
  "azurerm.function_app.ip_restriction.subnet_id": null,
  "azurerm.function_app.ip_restriction.virtual_network_subnet_id": null,
  "azurerm.function_app.ip_restriction.name": null,
  "azurerm.function_app.ip_restriction.priority": null,
  "azurerm.function_app.ip_restriction.action": null,
  "azurerm.function_app.scm_ip_restriction.ip_address": null,
  "azurerm.function_app.scm_ip_restriction.service_tag": null,
  "azurerm.function_app.scm_ip_restriction.subnet_id": null,
  "azurerm.function_app.scm_ip_restriction.virtual_network_subnet_id": null,
  "azurerm.function_app.scm_ip_restriction.name": null,
  "azurerm.function_app.scm_ip_restriction.priority": null,
  "azurerm.function_app.scm_ip_restriction.action": null,

  "azurerm.api_management.policy.xml_content": null,
  "azurerm.api_management.policy.xml_link": null,
};
