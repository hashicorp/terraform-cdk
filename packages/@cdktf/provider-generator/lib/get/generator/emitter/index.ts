// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
export * from "./attributes-emitter";
export * from "./resource-emitter";
export * from "./struct-emitter";
