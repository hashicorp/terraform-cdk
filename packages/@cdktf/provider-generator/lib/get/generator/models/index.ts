// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
export * from "./attribute-model";
export * from "./attribute-type-model";
export * from "./resource-model";
export * from "./scope";
export * from "./struct";
