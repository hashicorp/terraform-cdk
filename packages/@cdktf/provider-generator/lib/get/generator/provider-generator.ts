// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { CodeMaker, toCamelCase } from "codemaker";
import { LANGUAGES, logger, TerraformProviderConstraint } from "@cdktf/commons";
import {
  FQPN,
  parseFQPN,
  ProviderName,
  ProviderSchema,
} from "./provider-schema";
import { ResourceModel } from "./models";
import { ResourceParser } from "./resource-parser";
import { ResourceEmitter, StructEmitter } from "./emitter";
import {
  ConstructsMakerProviderTarget,
  ConstructsMakerTarget,
} from "../constructs-maker";

interface ProviderData {
  name: string;
  source: string;
  version: string;
}

const isMatching = (
  target: ConstructsMakerTarget,
  terraformSchemaName: string
): boolean => {
  if (target.isModule) return false;

  const elements = terraformSchemaName.split("/");

  if (elements.length === 1) {
    return target.source === terraformSchemaName;
  } else {
    const [hostname, scope, provider] = elements;

    if (!hostname || !scope || !provider) {
      throw new Error(`can't handle ${terraformSchemaName}`);
    }

    return target.name === provider;
  }
};

export interface ProviderConstraints {
  [fqn: string]: ProviderData;
}

export class TerraformProviderGenerator {
  private resourceParser = new ResourceParser();
  private resourceEmitter: ResourceEmitter;
  private structEmitter: StructEmitter;
  public versions: { [fqpn: string]: string | undefined } = {};

  constructor(
    private readonly code: CodeMaker,
    private readonly schema: ProviderSchema
  ) {
    this.code.indentation = 2;
    this.resourceEmitter = new ResourceEmitter(this.code);
    this.structEmitter = new StructEmitter(this.code);
  }

  private getProviderByConstraint(
    providerConstraint: ConstructsMakerTarget
  ): FQPN | undefined {
    return Object.keys(this.schema.provider_schemas || {}).find((fqpn) =>
      isMatching(providerConstraint, fqpn)
    ) as FQPN | undefined;
  }

  public generate(providerConstraint: ConstructsMakerTarget) {
    const fqpn = this.getProviderByConstraint(providerConstraint);
    if (!fqpn) {
      logger.debug(
        `Could not find provider constraint for ${providerConstraint} in schema: ${JSON.stringify(
          this.schema,
          null,
          2
        )}`
      );
      throw new Error(
        `Could not find provider with constraint ${providerConstraint}`
      );
    }

    const providerVersion = this.schema.provider_versions?.[fqpn];
    this.emitProvider(fqpn, providerVersion, providerConstraint);
    this.versions[fqpn] = providerVersion;
  }

  public generateAll() {
    for (const fqpn of Object.keys(this.schema.provider_schemas || {})) {
      this.generate(
        new ConstructsMakerProviderTarget(
          new TerraformProviderConstraint(fqpn),
          LANGUAGES[0]
        )
      );
    }
  }

  public async save(outdir: string) {
    await this.code.save(outdir);
  }

  public buildResourceModels(fqpn: FQPN): ResourceModel[] {
    const provider = this.schema.provider_schemas?.[fqpn];
    if (!provider) {
      throw new Error(`Can not find provider '${fqpn}' in schema`);
    }

    const resources = Object.entries(provider.resource_schemas || {}).map(
      ([type, resource]) =>
        this.resourceParser.parse(fqpn, type, resource, "resource")
    );

    const dataSources = Object.entries(provider.data_source_schemas || {}).map(
      ([type, resource]) =>
        this.resourceParser.parse(fqpn, `data_${type}`, resource, "data_source")
    );

    return ([] as ResourceModel[]).concat(...resources, ...dataSources);
  }

  public getClassNameForResource(terraformType: string) {
    return this.resourceParser.getClassNameForResource(terraformType);
  }

  public getNamespaceNameForResource(terraformType: string) {
    return this.resourceParser.getNamespaceNameForResource(terraformType);
  }

  private emitProvider(
    fqpn: FQPN,
    providerVersion?: string,
    constraint?: ConstructsMakerTarget
  ) {
    const { name } = parseFQPN(fqpn);
    const provider = this.schema.provider_schemas?.[fqpn];
    if (!provider) {
      throw new Error(`Can not find provider '${fqpn}' in schema`);
    }

    const files: string[] = [];
    this.buildResourceModels(fqpn).forEach((resourceModel) => {
      if (constraint) {
        resourceModel.providerVersionConstraint = constraint.version;
        resourceModel.terraformProviderSource = constraint.source;
      }
      resourceModel.providerVersion = providerVersion;

      if (resourceModel.structsRequireSharding) {
        files.push(this.emitResourceWithComplexStruct(resourceModel));
      } else {
        files.push(this.emitResource(resourceModel));
      }
      this.emitResourceReadme(resourceModel);
    });

    if (provider.provider) {
      const providerResource = this.resourceParser.parse(
        fqpn,
        `provider`,
        provider.provider,
        "provider"
      );
      if (constraint) {
        providerResource.providerVersionConstraint = constraint.version;
        providerResource.terraformProviderSource = constraint.source;
      }
      providerResource.providerVersion = providerVersion;
      files.push(this.emitResource(providerResource));
      this.emitResourceReadme(providerResource);
    }

    this.emitIndexFile(name, files);
    this.emitLazyIndexFile(name, files);
  }

  private emitResourceReadme(resource: ResourceModel): void {
    const filePath = `${resource.namespaceFolderPath}/README.md`;
    this.code.openFile(filePath);
    this.code.line(`# \`${resource.terraformType}\``);
    this.code.line();
    const type = resource.isProvider
      ? resource.provider
      : resource.terraformType;
    this.code.line(
      `Refer to the Terraform Registory for docs: [\`${type}\`](${resource.linkToDocs}).`
    );
    this.code.closeFile(filePath);
  }

  private emitIndexFile(provider: ProviderName, files: string[]): void {
    const folder = `providers/${provider}`;
    const filePath = `${folder}/index.ts`;
    this.code.openFile(filePath);
    this.code.line("// generated by cdktf get");
    for (const file of files) {
      const dirName = file.replace(`${folder}/`, "").replace("/index.ts", "");
      this.code.line(
        `export * as ${toCamelCase(dirName)} from './${dirName}';`
      );
    }
    this.code.line();
    this.code.closeFile(filePath);
  }

  private emitLazyIndexFile(provider: ProviderName, files: string[]): void {
    const folder = `providers/${provider}`;
    const filePath = `${folder}/lazy-index.ts`;
    this.code.openFile(filePath);
    this.code.line("// generated by cdktf get");
    for (const file of files) {
      const dirName = file.replace(`${folder}/`, "").replace("/index.ts", "");
      this.code.line(
        `Object.defineProperty(exports, '${toCamelCase(
          dirName
        )}', { get: function () { return require('./${dirName}'); } });`
      );
    }
    this.code.line();
    this.code.closeFile(filePath);
  }

  private emitResource(resource: ResourceModel): string {
    this.code.openFile(resource.filePath);
    this.emitFileHeader(resource);
    this.structEmitter.emit(resource);
    this.resourceEmitter.emit(resource);
    this.code.closeFile(resource.filePath);

    return resource.filePath;
  }

  private emitResourceWithComplexStruct(resource: ResourceModel) {
    const generatedFiles = [];

    // drop the last segment of the filepath
    const filePath = resource.filePath;
    this.code.openFile(filePath);
    this.code.line(`// generated from terraform resource schema`);
    this.code.line();

    if (resource.structsRequireSharding) {
      if (resource.referencedTypes.length > 0) {
        this.code.line(
          `import { ${resource.referencedTypes.join(", \n")}} from './${
            resource.structsFolderName
          }'`
        );
      }

      this.code.line(`export * from './${resource.structsFolderName}'`);

      resource.importStatements.forEach((statement) =>
        this.code.line(statement)
      );

      this.structEmitter.emitInterface(resource, resource.configStruct);
      this.resourceEmitter.emit(resource);

      this.code.closeFile(filePath);

      this.structEmitter.emit(resource);
      generatedFiles.push(resource.fileName);
      generatedFiles.push(resource.structsFolderName);
    } else {
      resource.importStatements.forEach((statement) =>
        this.code.line(statement)
      );

      this.structEmitter.emit(resource);
      this.resourceEmitter.emit(resource);
      this.code.closeFile(filePath);
      generatedFiles.push(resource.fileName);
    }

    return filePath;
  }

  private emitFileHeader(resource: ResourceModel) {
    this.code.line(`// ${resource.linkToDocs}`);
    this.code.line(`// generated from terraform resource schema`);
    this.code.line();
    resource.importStatements.forEach((statement) => this.code.line(statement));
    this.code.line();
    this.code.line("// Configuration");
    this.code.line();
  }
}
