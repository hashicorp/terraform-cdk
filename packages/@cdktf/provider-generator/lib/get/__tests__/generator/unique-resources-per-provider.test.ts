import * as fs from "fs";
import * as os from "os";
import * as path from "path";
import { TerraformProviderGenerator } from "../../generator/provider-generator";
import { CodeMaker } from "codemaker";

test("two providers with same resource name do not collide", async () => {
  const code = new CodeMaker();
  const workdir = fs.mkdtempSync(
    path.join(os.tmpdir(), "unique-resources-per-provider.test")
  );
  console.log("workdir", workdir);
  const spec = JSON.parse(
    fs.readFileSync(
      path.join(
        __dirname,
        "fixtures",
        "unique-resources-per-provider.text.fixture.json"
      ),
      "utf-8"
    )
  );
  new TerraformProviderGenerator(code, spec);
  await code.save(workdir);

  const providerOneOutput = fs.readFileSync(
    path.join(workdir, "providers/provider_one/resource-name.ts"),
    "utf-8"
  );
  expect(providerOneOutput).toMatchSnapshot();

  const providerTwoOutput = fs.readFileSync(
    path.join(workdir, "providers/provider_two/resource-name.ts"),
    "utf-8"
  );
  expect(providerTwoOutput).toMatchSnapshot();
});
