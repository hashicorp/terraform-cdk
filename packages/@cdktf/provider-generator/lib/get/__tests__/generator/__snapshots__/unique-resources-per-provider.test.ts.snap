// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`two providers with same resource name do not collide 1`] = `
"// https://www.terraform.io/docs/providers/provider_one/r/resource_name
// generated from terraform resource schema

import { Construct } from 'constructs';
import * as cdktf from 'cdktf';

// Configuration

export interface ResourceNameConfig extends cdktf.TerraformMetaArguments {
  /**
  * resource number 1
  * 
  * Docs at Terraform Registry: {@link https://www.terraform.io/docs/providers/provider_one/r/resource_name#attribute_one ResourceName#attribute_one}
  */
  readonly attributeOne: boolean | cdktf.IResolvable;
}

/**
* Represents a {@link https://www.terraform.io/docs/providers/provider_one/r/resource_name resource_name}
*/
export class ResourceName extends cdktf.TerraformResource {

  // =================
  // STATIC PROPERTIES
  // =================
  public static readonly tfResourceType: string = \\"resource_name\\";

  // ===========
  // INITIALIZER
  // ===========

  /**
  * Create a new {@link https://www.terraform.io/docs/providers/provider_one/r/resource_name resource_name} Resource
  *
  * @param scope The scope in which to define this construct
  * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
  * @param options ResourceNameConfig
  */
  public constructor(scope: Construct, id: string, config: ResourceNameConfig) {
    super(scope, id, {
      terraformResourceType: 'resource_name',
      terraformGeneratorMetadata: {
        providerName: 'provider_one'
      },
      provider: config.provider,
      dependsOn: config.dependsOn,
      count: config.count,
      lifecycle: config.lifecycle
    });
    this._attributeOne = config.attributeOne;
  }

  // ==========
  // ATTRIBUTES
  // ==========

  // attribute_one - computed: false, optional: false, required: true
  private _attributeOne?: boolean | cdktf.IResolvable; 
  public get attributeOne() {
    return this.getBooleanAttribute('attribute_one');
  }
  public set attributeOne(value: boolean | cdktf.IResolvable) {
    this._attributeOne = value;
  }
  // Temporarily expose input value. Use with caution.
  public get attributeOneInput() {
    return this._attributeOne;
  }

  // =========
  // SYNTHESIS
  // =========

  protected synthesizeAttributes(): { [name: string]: any } {
    return {
      attribute_one: cdktf.booleanToTerraform(this._attributeOne),
    };
  }
}
"
`;

exports[`two providers with same resource name do not collide 2`] = `
"// https://www.terraform.io/docs/providers/provider_two/r/resource_name
// generated from terraform resource schema

import { Construct } from 'constructs';
import * as cdktf from 'cdktf';

// Configuration

export interface ResourceNameConfig extends cdktf.TerraformMetaArguments {
  /**
  * resource number 1
  * 
  * Docs at Terraform Registry: {@link https://www.terraform.io/docs/providers/provider_two/r/resource_name#attribute_one ResourceName#attribute_one}
  */
  readonly attributeOne: boolean | cdktf.IResolvable;
}

/**
* Represents a {@link https://www.terraform.io/docs/providers/provider_two/r/resource_name resource_name}
*/
export class ResourceName extends cdktf.TerraformResource {

  // =================
  // STATIC PROPERTIES
  // =================
  public static readonly tfResourceType: string = \\"resource_name\\";

  // ===========
  // INITIALIZER
  // ===========

  /**
  * Create a new {@link https://www.terraform.io/docs/providers/provider_two/r/resource_name resource_name} Resource
  *
  * @param scope The scope in which to define this construct
  * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
  * @param options ResourceNameConfig
  */
  public constructor(scope: Construct, id: string, config: ResourceNameConfig) {
    super(scope, id, {
      terraformResourceType: 'resource_name',
      terraformGeneratorMetadata: {
        providerName: 'provider_two'
      },
      provider: config.provider,
      dependsOn: config.dependsOn,
      count: config.count,
      lifecycle: config.lifecycle
    });
    this._attributeOne = config.attributeOne;
  }

  // ==========
  // ATTRIBUTES
  // ==========

  // attribute_one - computed: false, optional: false, required: true
  private _attributeOne?: boolean | cdktf.IResolvable; 
  public get attributeOne() {
    return this.getBooleanAttribute('attribute_one');
  }
  public set attributeOne(value: boolean | cdktf.IResolvable) {
    this._attributeOne = value;
  }
  // Temporarily expose input value. Use with caution.
  public get attributeOneInput() {
    return this._attributeOne;
  }

  // =========
  // SYNTHESIS
  // =========

  protected synthesizeAttributes(): { [name: string]: any } {
    return {
      attribute_one: cdktf.booleanToTerraform(this._attributeOne),
    };
  }
}
"
`;
